dmoj/cptbox/compiler_isolate.py

```diff
@@ -103,6 +148,47 @@ def __init__(self, *, tmpdir, read_fs, write_fs):
                 }
             )
 
+    def fullpath_from_reg_and_dirfd(self, debugger: Debugger, *, file_reg, dirfd):
+        rel_file = self.get_rel_file(debugger, reg=file_reg)
+        return self.get_full_path_unnormalized(debugger, rel_file, dirfd=dirfd)
+
+    @syscall_simulator
+    def simulate_link(self, debugger: Debugger) -> None:
+        dst = self.fullpath_from_reg_and_dirfd(debugger, file_reg=1, dirfd=AT_FDCWD)
+        self._access_check(debugger, dst, self.write_fs_jail)
+        src = self.fullpath_from_reg_and_dirfd(debugger, file_reg=0, dirfd=AT_FDCWD)
+        os.link(src, dst)
+
+    @syscall_simulator
+    def simulate_linkat(self, debugger: Debugger) -> None:
```

Comment:
I'd feel a lot more confident about these *at calls if we validated that the `dirfd` is something we know about, and disallowed otherwise. It's easy to imagine allowing a call we shouldn't have because we got the `at` part wrong. This was less scary when the most that could happen was reads.

Reply:
Sorry, I mean we should probably be validating the flags register.

Reply:
So, throw the dirfd through the jail as well?

Reply:
Like, read the path given by `dirfd`, and validate that they have at least read perms on it, and then check that the full path has write perms?

Reply:
No, I mean that for instance `linkat` has a flag register that can specify `AT_SYMLINK_FOLLOW`, which changes how the path should be interpreted. We don't implement support for this flag. As long as that is true, we shouldn't allow that flag. Failing with `EINVAL` if flags is nonzero is safe.

Reply:
Do we need to do this for generic checking too (like in handle_renameat), and if so, for which calls?

Reply:
Added flag checking for `linkat`, since that is the only call with flags that we simulate. `renameat` does not have flags; `renameat2` does.